File: from-2601-to-2700/2684-maximum-number-of-moves-in-a-grid/solution.go
type Cell struct {
    row int
    col int
}

func maxMoves(grid [][]int) int {
    m, n := len(grid), len(grid[0])
    memo := make(map[Cell]int)
    dirs := [][]int{[]int{0, 1}, []int{-1, 1}, []int{1, 1}}

    var dp func (r, c int) int
    dp = func (r, c int) int {
        cell := Cell{row: r, col: c}
        if v, ok := memo[cell]; ok {
            return v
        }
        maxPath := 0
        for _, dir := range dirs {
            nr, nc := r + dir[0], c + dir[1]
            if nr >= 0 && nr < m && nc >= 0 && nc < n && grid[nr][nc] > grid[r][c] {
                maxPath = max(maxPath, 1 + dp(nr, nc))
            }
        }
        memo[cell] = maxPath
        return maxPath
    }
    result := 0
    for r := 0; r < m; r ++ {
        result = max(result, dp(r, 0))
    }
    return result
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: from-2601-to-2700/2684-maximum-number-of-moves-in-a-grid/solution.py
class Solution:
    def maxMoves(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        dirs = [(0, 1), (1, 1), (-1, 1)]
        cache = dict()

        def dfs(r: int, c: int) -> int:
            if (r, c) in cache:
                return cache[(r, c)]
            result = 0
            for dr, dc in dirs:
                new_r, new_c = r + dr, c + dc
                if 0 <= new_r < m and new_c < n and grid[r][c] < grid[new_r][new_c]:
                    result = max(result, 1 + dfs(new_r, new_c))

            cache[(r, c)] = result
            return result

        return max(dfs(row, 0) for row in range(m))